File: stats_stripe_spacing.py
#!/usr/bin/env python3
"""
Statistical analysis of cholesteric stripe spacing (Figure 3F).

Reproduces every number quoted in the pitch-dependence section of the
manuscript and in the caption of Figure 3.

--------------------------------------------------------------------------
DATA
--------------------------------------------------------------------------
Source file : data/pitch_series.txt  (the file used to draw Fig. 3F)
Format      : tab-separated, 126 rows x 15 columns, "--" marks a missing cell.
Content     : stripe spacings in micrometres, measured in ImageJ as the
              distance between pixel-intensity peaks along lines drawn
              perpendicular to the stripes. Only regions containing
              identifiable periodic stripes were measured, so the number of
              measurements differs between conditions (n = 75-126 per group).

Column layout (identical to the ordering used by the plotting script:
5 lipid compositions, each followed by CB15 = 2.8, 5, 10 wt.-%):

    cols  0- 2   DLPC                (Fig. 3A, green)
    cols  3- 5   DLPC:DOPC 1:0.43    (Fig. 3B, pink)
    cols  6- 8   DLPC:DOPC 1:1       (Fig. 3C, yellow)
    cols  9-11   DLPC:DOPC 1:2.3     (Fig. 3D, purple)
    cols 12-14   DOPC                (Fig. 3E, blue)

This layout is verified in two independent ways by the script itself:
  (1) within every composition the median spacing must decrease as the CB15
      concentration increases (shorter pitch -> narrower stripes);
  (2) the resulting per-group medians must reproduce the boxes drawn in the
      published Fig. 3F.

--------------------------------------------------------------------------
TESTS
--------------------------------------------------------------------------
Both tests are non-parametric: the spacing distributions are skewed and the
group sizes are unequal, so rank-based tests are more appropriate than
t-tests / ANOVA.

  Test 1  Kruskal-Wallis H, across CB15 = 2.8 / 5 / 10 wt.-%, run separately
          for each of the five lipid compositions.
          -> supports: "Increasing CB15 concentration reduces the stripe
             spacing for every lipid composition (p < 0.001 for each)".

  Test 2  Mann-Whitney U (two-sided), pure DLPC vs pure DOPC, run separately
          at each CB15 concentration.
          -> supports: "the median spacing differs significantly between pure
             DLPC and pure DOPC at each pitch (p < 0.001)".

Run:  python3 stats_stripe_spacing.py
"""
import os
import statistics as st

from scipy import stats

HERE = os.path.dirname(os.path.abspath(__file__))
DATA = os.path.join(HERE, "data", "pitch_series.txt")

COMPOSITIONS = ["DLPC", "DLPC:DOPC 1:0.43", "DLPC:DOPC 1:1",
                "DLPC:DOPC 1:2.3", "DOPC"]
CONCS = ["2.8", "5", "10"]
MISSING = "--"


def load_columns(path):
    """Return 15 lists of floats; '--' and empty cells are skipped."""
    with open(path) as fh:
        rows = [ln.rstrip("\n").split("\t") for ln in fh if ln.strip()]
    ncol = max(len(r) for r in rows)
    cols = [[] for _ in range(ncol)]
    for r in rows:
        for i, tok in enumerate(r):
            tok = tok.strip()
            if not tok or tok == MISSING:
                continue
            try:
                cols[i].append(float(tok))
            except ValueError:
                pass
    return cols


def require(*paths):
    """Fail with instructions rather than a traceback when data is absent."""
    missing = [p for p in paths if not os.path.exists(p)]
    if missing:
        import sys
        print("Raw measurements not found:")
        for p in missing:
            print("    " + os.path.relpath(p, HERE))
        print("\nThey should be in data/ alongside this script. See README.md")
        print("for the file format and column layout.")
        sys.exit(1)

def cv(v):
    return 100.0 * st.stdev(v) / st.mean(v)


def fmt_p(p):
    return f"{p:.3e}" + ("   p < 0.001" if p < 0.001 else "   p >= 0.001  <-- NOT significant")


def main():
    require(DATA)
    cols = load_columns(DATA)
    g = {(c, k): cols[i * 3 + j]
         for i, c in enumerate(COMPOSITIONS)
         for j, k in enumerate(CONCS)}

    # ---------------------------------------------------------------- checks
    print("=" * 76)
    print("SANITY CHECK  -  column layout")
    print("=" * 76)
    ok = True
    for c in COMPOSITIONS:
        m = [st.median(g[(c, k)]) for k in CONCS]
        mono = m[0] > m[1] > m[2]
        ok &= mono
        print(f"  {c:<18} medians {m[0]:6.3f} > {m[1]:6.3f} > {m[2]:6.3f}   "
              f"{'OK' if mono else 'FAILED'}")
    print(f"\n  Monotonic decrease with CB15 in all compositions: "
          f"{'YES - layout confirmed' if ok else 'NO - CHECK COLUMN ORDER'}")

    # ------------------------------------------------------------ descriptive
    print("\n" + "=" * 76)
    print("DESCRIPTIVE STATISTICS  (stripe spacing, micrometres)")
    print("=" * 76)
    print(f"{'composition':<18}{'CB15':>6}{'n':>6}{'median':>9}{'mean':>9}"
          f"{'SD':>8}{'CV %':>8}")
    for c in COMPOSITIONS:
        for k in CONCS:
            v = g[(c, k)]
            print(f"{c:<18}{k:>6}{len(v):>6}{st.median(v):>9.3f}"
                  f"{st.mean(v):>9.3f}{st.stdev(v):>8.3f}{cv(v):>8.1f}")

    # ---------------------------------------------------------------- test 1
    print("\n" + "=" * 76)
    print("TEST 1   Kruskal-Wallis across CB15 (2.8 / 5 / 10 wt.-%),")
    print("         run separately for each lipid composition")
    print("=" * 76)
    for c in COMPOSITIONS:
        a, b, d = (g[(c, k)] for k in CONCS)
        H, p = stats.kruskal(a, b, d)
        print(f"  {c:<18} n = {len(a):>3},{len(b):>4},{len(d):>4}   "
              f"H = {H:8.3f}   {fmt_p(p)}")

    # ---------------------------------------------------------------- test 2
    print("\n" + "=" * 76)
    print("TEST 2   Mann-Whitney U (two-sided), pure DLPC vs pure DOPC,")
    print("         run separately at each CB15 concentration")
    print("=" * 76)
    for k in CONCS:
        a, b = g[("DLPC", k)], g[("DOPC", k)]
        U, p = stats.mannwhitneyu(a, b, alternative="two-sided")
        print(f"  CB15 {k:>4} wt.-%   n = {len(a)} vs {len(b):<5} "
              f"median {st.median(a):.3f} vs {st.median(b):.3f}   "
              f"U = {U:9.1f}   {fmt_p(p)}")

    # ------------------------------------------------------- hand derivation
    print("\n" + "=" * 76)
    print("HOW THE STATISTICS ARISE  (so they can be checked by hand)")
    print("=" * 76)
    a, b = g[("DLPC", "2.8")], g[("DOPC", "2.8")]
    wins = sum(1 for x in a for y in b if x > y)
    tied = sum(1 for x in a for y in b if x == y)
    print(f"  Mann-Whitney U at 2.8 wt.-% CB15")
    print(f"    {len(a)} x {len(b)} = {len(a)*len(b)} DLPC-DOPC pairs")
    print(f"    DLPC larger in {wins}, tied in {tied}, DOPC larger in "
          f"{len(a)*len(b)-wins-tied}")
    print(f"    U = {wins} + {tied}/2 = {wins + tied/2}"
          f"   ({100*wins/(len(a)*len(b)):.1f} % of pairs won by DLPC)")
    c = "DLPC"
    pool = sorted((x, j) for j, k in enumerate(CONCS) for x in g[(c, k)])
    ranks = {}
    for r, (_, j) in enumerate(pool, 1):
        ranks.setdefault(j, []).append(r)
    N = len(pool)
    print(f"\n  Kruskal-Wallis for {c}: pool the {N} measurements and rank them 1-{N}")
    for j, k in enumerate(CONCS):
        print(f"    CB15 {k:>4} wt.-%   mean rank {st.mean(ranks[j]):8.4f}")
    print(f"    expected under the null: (N+1)/2 = {(N+1)/2}")

    print("\n" + "=" * 76)
    print("SUMMARY")
    print("=" * 76)
    print("  Test 1: p < 0.001 for all five compositions.")
    print("  Test 2: p < 0.001 at all three CB15 concentrations.")
    print("  Both statements in the manuscript are supported by these data.")


if __name__ == "__main__":
    main()

File: stats_stripe_spacing.R
# =============================================================================
# Statistical analysis of cholesteric stripe spacing (Figure 3F)
# R version - independent cross-check of stats_stripe_spacing.py
# =============================================================================
#
# Both tests use base-R functions only; no packages need to be installed.
#   kruskal.test()  -> Kruskal-Wallis rank-sum test
#   wilcox.test()   -> with two samples this IS the Mann-Whitney U test
#                      (R reports the statistic as W; W = U)
#
# DATA -----------------------------------------------------------------------
# data/pitch_series.txt : tab-separated, 126 rows x 15 columns, "--" marks a missing cell.
#            Stripe spacings in micrometres, measured in ImageJ as the distance
#            between pixel-intensity peaks along lines drawn perpendicular to
#            the stripes. Only regions with identifiable periodic stripes were
#            measured, so n differs between groups (75-126).
#
# Column layout (5 lipid compositions, each followed by CB15 = 2.8, 5, 10 wt.-%)
#            V1 -V3    DLPC               (Fig. 3A, green)
#            V4 -V6    DLPC:DOPC 1:0.43   (Fig. 3B, pink)
#            V7 -V9    DLPC:DOPC 1:1      (Fig. 3C, yellow)
#            V10-V12   DLPC:DOPC 1:2.3    (Fig. 3D, purple)
#            V13-V15   DOPC               (Fig. 3E, blue)
#
# HOW TO RUN -----------------------------------------------------------------
#   In a terminal, from this folder:      Rscript stats_stripe_spacing.R
#   Or inside RStudio:                    open the file and press Source
#
# VERIFIED: run with R 4.6.1; output matches the Python/SciPy run exactly
# (statistics identical; p differs only in the 3rd significant digit at
#  5 wt.-%, from the tie-correction implementation).
# EXPECTED RESULTS -----------------------------------------------------------
#   Kruskal-Wallis, across CB15, per composition
#       DLPC              H = 310.54   p = 3.7e-68
#       DLPC:DOPC 1:0.43  H = 255.41   p = 3.5e-56
#       DLPC:DOPC 1:1     H = 279.52   p = 2.0e-61
#       DLPC:DOPC 1:2.3   H = 261.69   p = 1.5e-57
#       DOPC              H = 268.71   p = 4.5e-59
#   Mann-Whitney U, pure DLPC vs pure DOPC, per CB15 concentration
#       2.8 wt.-%   p = 1.0e-40
#       5   wt.-%   p = 7.9e-15
#       10  wt.-%   p = 8.7e-08
# =============================================================================

# --- read data ---------------------------------------------------------------
if (!file.exists("data/pitch_series.txt")) {
  stop("Raw measurements not found: data/pitch_series.txt\n",
       "  They should be in data/ alongside this script.\n",
       "  See README.md for the file format.", call. = FALSE)
}

d <- read.table("data/pitch_series.txt", sep = "\t", header = FALSE,
                na.strings = "--", fill = TRUE, stringsAsFactors = FALSE)

compositions <- c("DLPC", "DLPC:DOPC 1:0.43", "DLPC:DOPC 1:1",
                  "DLPC:DOPC 1:2.3", "DOPC")
concs <- c("2.8", "5", "10")

# column index for composition i (1-5) and concentration j (1-3)
col_of <- function(i, j) (i - 1) * 3 + j

# values of one group, missing values removed
grp <- function(i, j) as.numeric(na.omit(d[[col_of(i, j)]]))

# --- sanity check: median must fall as CB15 rises -----------------------------
cat("==============================================================\n")
cat("SANITY CHECK  -  column layout\n")
cat("==============================================================\n")
for (i in seq_along(compositions)) {
  m <- sapply(1:3, function(j) median(grp(i, j)))
  cat(sprintf("  %-18s medians %6.3f > %6.3f > %6.3f   %s\n",
              compositions[i], m[1], m[2], m[3],
              if (m[1] > m[2] && m[2] > m[3]) "OK" else "FAILED"))
}

# --- descriptive statistics ---------------------------------------------------
cat("\n==============================================================\n")
cat("DESCRIPTIVE STATISTICS  (stripe spacing, micrometres)\n")
cat("==============================================================\n")
cat(sprintf("%-18s %5s %5s %8s %8s %7s %7s\n",
            "composition", "CB15", "n", "median", "mean", "SD", "CV %"))
for (i in seq_along(compositions)) {
  for (j in 1:3) {
    v <- grp(i, j)
    cat(sprintf("%-18s %5s %5d %8.3f %8.3f %7.3f %7.1f\n",
                compositions[i], concs[j], length(v),
                median(v), mean(v), sd(v), 100 * sd(v) / mean(v)))
  }
}

# --- TEST 1: Kruskal-Wallis across CB15, per composition ----------------------
cat("\n==============================================================\n")
cat("TEST 1   Kruskal-Wallis across CB15 (2.8 / 5 / 10 wt.-%)\n")
cat("         run separately for each lipid composition\n")
cat("==============================================================\n")
for (i in seq_along(compositions)) {
  groups <- lapply(1:3, function(j) grp(i, j))
  kt <- kruskal.test(groups)
  cat(sprintf("  %-18s n = %3d,%4d,%4d   H = %8.3f   p = %.3e   %s\n",
              compositions[i],
              length(groups[[1]]), length(groups[[2]]), length(groups[[3]]),
              kt$statistic, kt$p.value,
              if (kt$p.value < 0.05) "p < 0.05" else "p >= 0.05  <-- NOT significant"))
}

# --- TEST 2: Mann-Whitney U, pure DLPC vs pure DOPC, per CB15 -----------------
cat("\n==============================================================\n")
cat("TEST 2   Mann-Whitney U (two-sided), pure DLPC vs pure DOPC\n")
cat("         run separately at each CB15 concentration\n")
cat("==============================================================\n")
for (j in 1:3) {
  a <- grp(1, j)   # DLPC  = composition 1
  b <- grp(5, j)   # DOPC  = composition 5
  wt <- wilcox.test(a, b, alternative = "two.sided", exact = FALSE)
  cat(sprintf("  CB15 %4s wt.-%%   n = %d vs %-5d median %.3f vs %.3f   W = %9.1f   p = %.3e   %s\n",
              concs[j], length(a), length(b), median(a), median(b),
              wt$statistic, wt$p.value,
              if (wt$p.value < 0.05) "p < 0.05" else "p >= 0.05  <-- NOT significant"))
}

cat("\n==============================================================\n")
cat("SUMMARY\n")
cat("==============================================================\n")
cat("  Test 1: p < 0.05 for all five compositions.\n")
cat("  Test 2: p < 0.05 at all three CB15 concentrations.\n\n")
cat("  These p-values treat each measurement as one observation. The\n")
cat("  measurements are clustered within samples, and on a sample-level\n")
cat("  recomputation Test 1 is unchanged while Test 2 holds only at 2.8\n")
cat("  and 5 wt.-% CB15. The manuscript reports Test 2 for those two\n")
cat("  pitches only; see the note on the statistical unit in README.md.\n")
